### PR TITLE
Use the DSi System Menu Font (fixes #13, fixes #79)

### LINKED ISSUE #13: Games with Special Characters in the Title Look Glitchy
I'm running SRLoader on my Acekard 2i (as I cannot use TWLoader due to [this bug](https://github.com/Robz8/TWLoader/issues/347)). All of the Pokémon games (and I assume any others with special characters display incorrectly on both themes. On the DSi Menu theme. the é character seems to be replaced with the textures for the L/R button that display on the top screen (this also pushes the text outside of the box). On the Aura theme, the é is replaced with a two character wide black box that extends downwards about 5-6 lines worth of text.

### LINKED ISSUE #79: [Bug] foreign region icon string prints garbage
For example of Japanese rom (Which doesn't have English string at all) It'll print out garbage (Possibly ref to #13)

Japanese/All foreign string is encoded with UTF-16LE.

### PR DESCRIPTION
**Pull request by RonnChyran**
#### What's new?

The DSi System Menu uses a font with the following character set for its UI

The font was dumped and reformatted with proper UV coordinates to work with DSiMenu++, including support for most of the above charset, bar some of the private use glyphs that couldn't fit into one texture. The small font texture only contains the ASCII, extended ASCII charset, Hiragana and Katakana charset in order to fit into a 512 by 128 texture.

In order to support UTF16 code points, changes were made to font handling to use a lookup table instead of converting to ASCII. In addition, since many UTF16 codepoints can not be represented with a single `char`, the presence of the ASCII control character `0x0F` (SHIFT IN) will cause the next 2 bytes in the character array to be interpreted as a single UTF16 codepoint (unsigned short int) in little-endian order. For example, to print the letter あ (U+3042):

```c
char message[] = { 0x0F, 0x42, 0x30 };
printLarge(..., message); // Will print あ with the large font
```

Issues #13 (and clone issue #79) will be fixed once this pull request is complete. Full support for characters found in the System Menu font, including support for the Latin-A extended character block (issues such as é will not require special case replacements).

Font transparency is also properly implemented with `#FF00FF` sprite sheets. 

This does not change font handling in R4 menu, and there are no plans to port over improvements to the R4 theme.

### What is fixed?

* Garbled text for games with Japanese and non ASCII characters now displays properly in the DSi menu theme.
* Garbled text if the user had non ASCII characters in their profile name now displays properly in the DSi menu theme.
* The profile name font now has alpha-blended edges to look better in the dark theme.
* Left-aligned the game title in the title dialog so long titles don't overlap the game icon

#140 was closed in favor of this approach with a restricted charset and single font texture for performance reasons. The textures, UV coordinates, and character maps were dumped and generated with the use of custom built tools, the (albeit poorly documented) source code of which is available [here](https://github.com/RonnChyran/TWLFontTools).

#### Where have you tested it?

Nintendo DSi with latest bootstrap, Nintendo 3DS with latest bootstrap.

*** 
#### Pull Request status
- [x]  This PR has been tested using latest devkitPro, devkitARM, and EasyGL2D.  

_(Do not edit after this point)_
- [ ]  This PR is fully documented.
- [ ]  This PR has been tested before sending.
- [ ]  This PR follows C style and convention.

### CODE REVIEW

**Comment by RonnChyran** [comment]
TODO: 
- [ ] Reduce small font charset to fit in a 128px texture
- [ ] Try to get the top name drawing with raw grit bitmap texture

**Comment by Stack3r** [comment]
I know this is a font project but are you able to find/add the full dsi menu music and sound effects? *excuse me if you can’t*

**Comment by RonnChyran** [comment]
I did some digging on this and the only conclusion I could come up with is that the DSi menu music is compiled into the ARM9 binary. Extracting it is way beyond my level of experience, unfortunately.

I did find the System Settings sound effects though, check [this out](https://onlinesequencer.net/892461). There is a sample in the same ROM which I assume is what they modulate to produce the background music, it's quite neat. Unfortunately it's not the launcher music, but the settings music.

**Comment by Stack3r** [comment]
Hmm that’s pretty cool, can you implement any sound effects to dsimenu if robz says it’s okay?

**Comment by RonnChyran** [comment]
I don't see a problem with the current sound effects though.

**Comment by RonnChyran** [comment]
@Stack3r I managed to extract the sound data from the System Menu title. I'll look into it, but not in this PR.

**Comment by RonnChyran** [comment]
just have to tweak the palettes a little and we should be good

**Comment by Stack3r** [comment]
Nice work! Will this fix the random texture glitches for banners and text?

**Comment by RonnChyran** [comment]
@Robz8 This should be ready for review now. This PR also includes #146, but let me know if you want me to rebase it on ced05290a8d0ffc873a4c9b399586902a50dcd37 instead.

**Comment by RonnChyran** [comment]
@Stack3r I don't know about the random glitches, but it should fix the garbled text if you had any Japanese DS games. Also the é in Pokémon should display properly as well.

**Comment by Robz8** [comment]
Chinese version of WarioWare touched only shows 1 non-english character.

**Comment by RonnChyran** [comment]
Chinese and Korean are not supported as part of this PR, the size and quantity of textures required to support them would probably require replacing EasyGL2D with a custom renderer, and that's if the fonts were dumpable from a Chinese or Korean DSi. The Chinese characters in `large_font` were left in for a potential Japanese translation (also the fact that they fit in one texture anyways) .

File: romsel_dsimenutheme/arm9/source/graphics/graphics.cpp [hunk 1]
```diff
@@ -1091,47 +1127,38 @@ void topBgLoad() {
...
-							val = bmpPal_topSmallFont[5+((PersonalData->theme)*16)];
-							break;
-						case 0xD294:
-							val = bmpPal_topSmallFont[6+((PersonalData->theme)*16)];
-							break;
-						case 0xDAD6:
-							val = bmpPal_topSmallFont[7+((PersonalData->theme)*16)];
+						case 0xC210:
+							// blend the colors with the background to make it look better.
+							val = alphablend(bmpPal_topSmallFont[2+((PersonalData->theme)*16)], bg, 64);
```
Comment (RonnChyran) [inline romsel_dsimenutheme/arm9/source/graphics/graphics.cpp]: If alpha is set to 32 here (`alphablend(bmpPal_topSmallFont[2+((PersonalData->theme)*16)], bg, 32);), it reduces the "bold" look, but then causes the edges to look too dark on the dark theme.

**Comment by RonnChyran** [comment]
@Robz8 anything holding this up to be merged?

**Comment by RonnChyran** [comment]
Which characters in particular? I've only noticed a slight issue with i and j, but reducing the width of i to fit neatly inside the j screws up the spacing with other letters.

image 
image

The UV coordinates were generated with data from the DSi system font file so it *should* be identical to the official launcher, although there may have been some off by one errors.

|Official|This PR|
|--|--|
|image|image|

**Comment by RonnChyran** [comment]
I saw some more egregious problems with the small font glyphs for `i` and `k`, so those should be 

|Before|Fixed|
|---|---|
|image|image image|

I also tightened up the kerning for the capital letters.

|Before|Fixed|
|---|---|
|image|image|

**Comment by RonnChyran** [comment]
ah, I hadn't noticed the issue with p and r so that should be fixed now as well.

image

**Comment by RonnChyran** [comment]
Somehow the 1 ended up full width, but that should be fixed now.
image
image

**Comment by RonnChyran** [comment]
@Robz8 any letters I missed left to kern?

**Comment by Robz8** [comment]
Sorry, haven't checked.

**Comment by Robz8** [comment]
Ok, was finally able to test this again.
One letter that still needs fixing is the uppercase R, but aside from that, it looks good.
I'll merge this, and fix the R.

**Pull request merged**